### ISSUE TITLE
RFC: Support A2PO with DeepSpeed ZeRO-3 via an explicit reference model

### ISSUE THREAD

**Issue by DimensionSTP** [huggingface/trl#6831]
### Feature request

A2PO currently creates its frozen initial-policy reference internally with:

```python
self.ref_model = create_reference_model(model)
```

`create_reference_model()` intentionally rejects DeepSpeed ZeRO-3 because it relies on `deepcopy(model)`, which cannot safely reconstruct the complete initial state of an already partitioned model.

I propose adding an optional explicit `ref_model` argument to `A2POTrainer` for full-model training:

```python
policy = AutoModelForCausalLM.from_pretrained(model_id)
reference = AutoModelForCausalLM.from_pretrained(model_id)

trainer = A2POTrainer(
    model=policy,
    ref_model=reference,
    reward_funcs=reward_func,
    args=args,
    train_dataset=dataset,
)
```

Proposed semantics:

- `ref_model=None`: preserve the current `create_reference_model(model)` behavior, including its ZeRO-3 guard.
- Explicit `ref_model`: skip the deepcopy path, freeze all reference parameters, and set the model to evaluation mode.
- Reject `ref_model is model`.
- Under ZeRO-3, prepare the reference through TRL's existing `prepare_deepspeed()` inference/evaluation path.
- Preserve the current ZeRO-1/2 and non-DeepSpeed behavior.

The explicit reference must represent the same initial policy as the trainable model. A2PO uses this immutable policy in both Stage 1 optimal-value estimation and Stage 2 reference log-probability computation.

### Motivation

A2PO cannot currently initialize under ZeRO-3, even when users can afford and explicitly instantiate a separate reference model. Removing the existing guard is not safe: deepcopying a partitioned policy could yield incomplete or rank-local reference state and silently change the algorithm.

This is related to the general reference-model/ZeRO-3 constraint discussed in #827 and #3022, but those reports concern PPO/OnlineDPO. A2PO has a distinct two-stage reference contract and currently exposes no explicit reference-model API.

The behavior remains present on current main (`3f738b90ed9de71c76f0593f575ad62851c30ba1`).

I tested a minimal prototype based on main with the existing guard retained.

CPU validation:

- explicit reference avoids only the deepcopy path while ZeRO-3 is enabled
- omitted reference still raises the existing guard
- supplied reference is frozen and in eval mode
- passing the policy itself as the reference is rejected
- complete A2PO test module: 12 passed

Two-rank DeepSpeed ZeRO-3 smoke test:

- BF16, no CPU/NVMe offload
- Stage 1 optimal-value estimation completed
- Stage 2 completed one optimizer step on both ranks
- policy max parameter difference: `0.10156250`
- reference max parameter difference: `0.00000000`

I suggest keeping these concerns out of the initial change:

- PEFT adapter-disable optimization
- continued-PEFT initial-adapter reference semantics
- callback/factory-based reference construction
- automatic reference model identifiers or loading kwargs

Questions for maintainers:

1. Is an explicit `ref_model` object the preferred minimal API for A2PO ZeRO-3 support?
2. Should A2PO use `prepare_deepspeed()` only for ZeRO-3 while preserving its current ZeRO-1/2 path?
3. Should PEFT reference semantics remain a separate follow-up proposal?

### Your contribution

I can submit the tested minimal implementation and unit tests after agreement on the public API. I have also prepared a two-rank ZeRO-3 smoke test that verifies the policy changes while the reference remains unchanged. AI assistance was used during investigation and drafting; I reviewed the implementation, reproduction, and results directly.

**Comment by DimensionSTP** [comment]
@qgallouedec, could you advise on the proposed API, or point me to the appropriate reviewer?

Current main still creates A2PO’s reference internally through create_reference_model(), so the ZeRO-3 restriction remains. The previously tested prototype accepts an explicit frozen reference while retaining the existing guard when none is supplied.

Would an optional ref_model argument, initially scoped to full-model training with PEFT semantics deferred, be an acceptable direction for a PR? I can refresh the prototype and its tests against current main once the API direction is agreed. Thanks!
